File: sdk/nodejs/src/models/job-response.ts
import { ExecutionPolicy, JobPriority, JobStatus, LeaseInfo } from './common';
//Response object for job information.

export interface JobResponse {
  //Unique job identifier.

  jobId: string;
  //Job type.

  jobType: string;
  //Job priority.

  priority: JobPriority;
  //Current job status.

  status: JobStatus;
  //Job payload data.

  payload?: Record<string, any>;
  //Execution policy.

  execution?: ExecutionPolicy;
  //Maximum retry attempts.

  maxAttempts?: number;
  //Current attempt number.

  attemptCount?: number;
  //Job creation timestamp (ISO string).

  createdAt?: string;
  //Scheduled execution timestamp (ISO string).

  scheduledAt?: string;
  //Job start timestamp (ISO string).

  startedAt?: string;
  //Job completion timestamp (ISO string).

  completedAt?: string;
  //Next retry timestamp (ISO string).

  nextRetryAt?: string;
  //Error message (if failed).

  errorMessage?: string;
  //Job result data.

  result?: Record<string, any>;
  //Job tags.

  tags?: Record<string, string>;
  //Callback URL.

  callbackUrl?: string;
  //Tenant ID.

  tenantId?: string;
  //Worker ID that processed the job.

  workerId?: string;
  //Lease information.

  lease?: LeaseInfo;
}
//Helper function to check if a job is running.

export function isJobRunning(job: JobResponse): boolean {
  return job.status === 'RUNNING';
}
//Helper function to check if a job is completed successfully.

export function isJobSuccessful(job: JobResponse): boolean {
  return job.status === 'SUCCESS';
}
//Helper function to check if a job failed.

export function isJobFailed(job: JobResponse): boolean {
  return job.status === 'FAILED' || job.status === 'DLQ';
}
//Helper function to check if a job is queued.

export function isJobQueued(job: JobResponse): boolean {
  return job.status === 'QUEUED';
}
//Helper function to check if a job is retrying.

export function isJobRetrying(job: JobResponse): boolean {
  return job.status === 'RETRY_SCHEDULED';
}
//Helper function to get job duration in milliseconds.

export function getJobDuration(job: JobResponse): number | null {
  if (!job.startedAt || !job.completedAt) {
    return null;
  }

  const start = new Date(job.startedAt).getTime();
  const end = new Date(job.completedAt).getTime();
  return end - start;
}
//Helper function to get job wait time in milliseconds.

export function getJobWaitTime(job: JobResponse): number | null {
  if (!job.createdAt || !job.startedAt) {
    return null;
  }

  const created = new Date(job.createdAt).getTime();
  const started = new Date(job.startedAt).getTime();
  return started - created;
}

File: sdk/java/src/main/java/com/learnerview/sab/model/JobResponse.java
package com.learnerview.sab.model;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.Instant;
import java.util.Map;

/**
 * Response object for job information.
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonInclude(JsonInclude.Include.NON_NULL)
public class JobResponse {
    /**
     * Unique job identifier.
     */
    private String jobId;

    /**
     * Job type.
     */
    private String jobType;

    /**
     * Job priority.
     */
    private JobPriority priority;

    /**
     * Current job status.
     */
    private JobStatus status;

    /**
     * Job payload data.
     */
    private Map<String, Object> payload;

    /**
     * Execution policy.
     */
    private ExecutionPolicy execution;

    /**
     * Maximum retry attempts.
     */
    private Integer maxAttempts;

    /**
     * Current attempt number.
     */
    private Integer attemptCount;

    /**
     * Job creation timestamp.
     */
    private Instant createdAt;

    /**
     * Scheduled execution timestamp.
     */
    private Instant scheduledAt;

    /**
     * Job start timestamp.
     */
    private Instant startedAt;

    /**
     * Job completion timestamp.
     */
    private Instant completedAt;

    /**
     * Next retry timestamp (if applicable).
     */
    private Instant nextRetryAt;

    /**
     * Error message (if failed).
     */
    private String errorMessage;

    /**
     * Job result data.
     */
    private Map<String, Object> result;

    /**
     * Job tags.
     */
    private Map<String, String> tags;

    /**
     * Callback URL.
     */
    private String callbackUrl;

    /**
     * Tenant ID.
     */
    private String tenantId;

    /**
     * Worker ID that processed the job.
     */
    private String workerId;

    /**
     * Lease information.
     */
    private LeaseInfo lease;

    /**
     * Execution policy information.
     */
    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    @JsonInclude(JsonInclude.Include.NON_NULL)
    public static class ExecutionPolicy {
        private String type;
        private String endpoint;
        private Integer timeoutSeconds;
        private String callbackUrl;
        private Map<String, String> headers;
    }

    /**
     * Lease information.
     */
    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    @JsonInclude(JsonInclude.Include.NON_NULL)
    public static class LeaseInfo {
        private String owner;
        private String token;
        private String expiresAt;
        private String visibleAt;
    }
}
